protocols/relay/src/behaviour/handler.rs

```diff
@@ -380,11 +383,18 @@ pub struct Handler {
     alive_lend_out_substreams: FuturesUnordered<oneshot::Receiver<()>>,
     /// Futures relaying data for circuit between two peers.
     circuits: Futures<(CircuitId, PeerId, Result<(), std::io::Error>)>,
+
+    stop_requested_streams: VecDeque<outbound_stop::StopCommand>,
```

Comment:
To expand on https://github.com/libp2p/rust-libp2p/issues/3268#issuecomment-1694381620, I was thinking of something like:

```rust
struct OutboundStreams<TInfo> {
	info: VecDeque<TInfo>,
	open_stream_results: VecDeque<Result<Stream, StreamUpgradeError<()>>>,

	// possibly more state ...
}

impl<TInfo> OutboundStreams<TInfo> {
	pub fn request_new_stream(&mut self, info: TInfo);

	pub fn on_new_stream(&mut self, stream: Stream);
	pub fn on_stream_error(&mut self, error: StreamUpgradeError<()>);
	
	pub fn poll(&mut self, cx: Context<'_>) -> Poll<Event<TInfo>>>;
}

enum Event<TInfo> {
	
	Stream(Result<(Stream, TInfo), Error<TInfo>),
	/// Instruct the `ConnectionHandler` to request a new stream.
	RequestNewStream
}

enum Error<TInfo> {
	Timeout(TInfo),
    	Unsupported(TInfo),
	Io((TInfo, io::Error)),
}
```

The user would store one of these in their `ConnectionHandler` in place of the above `VecDeque<outbound_stop::StopCommand>`. You then call the `on_new_stream` function in the `on_connection_event` callback and `poll` inside the `ConnectionHandler::poll`.

Have a play with the design and let me know what you think :)

Reply:
It looks interesting.
I think this is a must-have helping tool, and we should implement something like this. But let's do this in a separate issue. Otherwise, the current issue will go far behind the scope of the relay feature.
I'm ready to make this tool right after the closing of this PR.
WDYT?

Reply:
Sounds good! Thank you for your help :)

Reply:
Resolving this because otherwise we can't merge.

Reply:
I've recorded this in a follow-up issue: https://github.com/libp2p/rust-libp2p/issues/4510